### PR TITLE
snapdtool, cmd: switch to FIPS mode when needed

### PR DESCRIPTION
**Pull request by bboozzoo**
Switch to FIPS compliance mode when the system has booted with FIPS enabled. Specifically, set GOFIPS=1 to enforce FIPS mode at runtime, and set up environment such that the relevant openssl modules can be located.

### CODE REVIEW

File: build-aux/snap/snapcraft.yaml
Comment (bboozzoo) [inline build-aux/snap/snapcraft.yaml:1]: This was extracted cherr picked into a separate PR: https://github.com/snapcore/snapd/pull/14094

File: snapdtool/fips_linux.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+		logger.Debugf("FIPS bootstrap complete")
+
+		// if we reached this place, then the initialization was
+		// completed successfully and we can drop the environment
+		// variables, other processes which may be invoked by snapd will
+		// perform the initialization cycle on their own when needed
+		os.Unsetenv("GOFIPS")
+		os.Unsetenv("SNAPD_FIPS_BOOTSTRAP_DONE")
+		os.Unsetenv("OPENSSL_MODULES")
+		os.Unsetenv("GO_OPENSSL_VERSION_OVERRIDE")
```
Comment (valentindavid) [inline snapdtool/fips_linux.go:89]: I feel like we should verify that initialization was done. If for some reason, some library changes a NEEDED to a dlopen, then some initialization will be delayed. It will be a well hidden bug then.

Could we somehow query go's openssl binding for where the modules are load from for example at this point?

File: snapdtool/fips_linux.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+	env := append(os.Environ(), []string{
+		"SNAPD_FIPS_BOOTSTRAP_DONE=1",
+		// make FIPS mod required at runtime, if the module was not
+		// found or the setup is incorrect snapd will fail in a
+		// predictable way
+		"GOFIPS=1",
+	}...)
+
+	if mod != "" {
```
Comment (valentindavid) [inline snapdtool/fips_linux.go:130]: In what legitimate case are the libraries not found, but we still re-exec with `GOFIPS=1`? Do we get anything better than returning here? Do we get an error from go when it tries to use openssl? Maybe a comment explaining what happens in that case would be helpful.

**Review by valentindavid — COMMENTED** [review]
A few questions here.
Also I suppose we will need some spread tests in a follow up when we have #13517.

File: snapdtool/fips_linux.go (hunk 1 above)
Comment (bboozzoo) [reply snapdtool/fips_linux.go:89]: the way the toolchain bits work, with GOFIPS=1 there would be a panic during startup already, so we'd never reach this place, see https://github.com/microsoft/go/blob/77f45f8554027fe20fb411b40328192354e58703/patches/0004-Add-OpenSSL-crypto-backend.patch#L279

File: snapdtool/fips_linux.go (hunk 2 above)
Comment (bboozzoo) [reply snapdtool/fips_linux.go:130]: there will be an explicit panic upon startup, which is the expected behavior, i.e. when FIPS is enabled it either works or fails

File: cmd/snap-repair/main.go [hunk 3]
```diff
@@ -59,6 +59,11 @@ func init() {
 var errOnClassic = fmt.Errorf("cannot use snap-repair on a classic system")
 
 func main() {
+	if err := snapdtool.MaybeSetupFIPS(); err != nil {
+		fmt.Fprintf(os.Stderr, "cannot check or enable FIPS mode: %v", err)
+		os.Exit(1)
```
Comment (zyga) [inline cmd/snap-repair/main.go:64]: This should not fail on missing "snapd" symlink. In other words, repair should not depend on snapd snap being mounted and operational, as that's one of the things repair may be necessary to fix in the field.

File: snapdtool/fips_linux.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+		// variables, other processes which may be invoked by snapd will
+		// perform the initialization cycle on their own when needed
+		os.Unsetenv("GOFIPS")
+		os.Unsetenv("SNAPD_FIPS_BOOTSTRAP_DONE")
+		os.Unsetenv("OPENSSL_MODULES")
+		os.Unsetenv("GO_OPENSSL_VERSION_OVERRIDE")
+		return nil
+	}
+
+	snapdRev, err := osReadlink(snapdSnap)
```
Comment (zyga) [inline snapdtool/fips_linux.go:93]: As we've been reviewing this, it became apparent that snapd snap is now required for snap-repair to function. In case snapd itself fails (current is gone, the snap doesn'n mount) then snap-repair is broken.

I think this is a regression as compared to core16 without snapd.snap, which carried snap-repair in the base and it was an entirely standalone executable.

To be clear we now have two dependencies:
 - on snapd snap (to provide snap-repair)
 - on snapd snap (to provide FIPS libiraries in FIPS mode)

CC @pedronis for comment.

**Review by zyga — APPROVED** [review]
Most of the code is good but I'm very worried about the relationship between snap-repair, snapd snap and fips.

I'll leave a +1 with a inline request for comment from Samuele.

File: snapdtool/fips_linux.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+		// found or the setup is incorrect snapd will fail in a
+		// predictable way
+		"GOFIPS=1",
+	}...)
+
+	if mod != "" {
+		// version override uses the version suffix right after *.so.
+		libVer := strings.TrimPrefix(filepath.Ext(lib), ".")
+		logger.Debugf("found FIPS library and module at %s (ver %s) and %s", lib, libVer, mod)
+		// FIPS module not found, we still let the process reexec
```
Comment (alexmurray) [inline snapdtool/fips_linux.go:134]: hmm perhaps I am missing something but this comment doesn't seem to match the code - we have just done a debug print saying we found the FIPS library and module but this comment says FIPS module not found...?

**Review by alexmurray — CHANGES_REQUESTED** [review]
LGTM other than the outstanding comment from @zyga re wanting snap-repair to not exit when the snapd symlink doesn't exist - in that case it seems like you should check the returned err and not die in the case it returns EINVAL/ENOENT etc.

File: snapdtool/fips_linux.go (hunk 4 above)
Comment (pedronis) [reply snapdtool/fips_linux.go:93]: I think we need to reorganize things such that we depend on UC on /usr/lib/snapd  and not the snapd current link, but we need to analyze again how the former itself is used/changed

**Review by pedronis — COMMENTED** [review]
I made a comment

File: cmd/snap-repair/main.go (hunk 3 above)
Comment (pedronis) [reply cmd/snap-repair/main.go:64]: but I think maybe dynamic linking already fails in the new snapd snap if the symlink is not there? (cc @valentindavid @bboozzoo ) that why we added more code in various places to make sure that the symlink is there at all times, but then it might be the case though that we need to add a failure handler to the snap-repair service that tries to put the symlink in place in as much cases as possible?

Comment (pedronis) [reply cmd/snap-repair/main.go:64]: I answered the first question on my own:

```
$ ldd /snap/snapd/current/usr/lib/snapd/snap-failure 
	linux-vdso.so.1 (0x00007ffe213cd000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x0000703de6800000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x0000703de6a77000)
	/snap/snapd/current/usr/lib/x86_64-linux-gnu/ld-linux-x86-64.so.2 => /lib64/ld-linux-x86-64.so.2 (0x0000703de6a91000)
```

Comment (bboozzoo) [reply cmd/snap-repair/main.go:64]: I'll revert this change and leave a TODO instead which we can address once the uncertainty around snap-repair is resolved.

File: snapdtool/fips_linux.go (hunk 5 above)
Comment (bboozzoo) [reply snapdtool/fips_linux.go]: thanks for spotting, I forgot to drop the comment after some earlier tweaks I made

File: cmd/snap-repair/main.go (hunk 3 above)
Comment (bboozzoo) [reply cmd/snap-repair/main.go]: this is now done, I left a TODO

**Review by pedronis — COMMENTED** [review]
couple of comments

File: snapdtool/fips_linux.go [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+		if !strings.HasPrefix(exe, currentRevSnapdSnap+"/") {
+			// this is only supported for reexecing from the snapd snap
+			return nil
+		}
+	}
+
+	lib, mod := findFIPSLibsAndModules(currentRevSnapdSnap)
+
+	env := append(os.Environ(), []string{
+		"SNAPD_FIPS_BOOTSTRAP_DONE=1",
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply snapdtool/fips_linux.go]: Thanks, I've updated the name.

File: snapdtool/fips_linux.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+	snapdRev, err := osReadlink(snapdSnap)
+	if err != nil {
+		return err
+	}
+
+	currentRevSnapdSnap := filepath.Join(dirs.SnapMountDir, "snapd", snapdRev)
+
+	logger.Debugf("snapd snap: %s", currentRevSnapdSnap)
+
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply snapdtool/fips_linux.go:101]: I played with this a little, and tried a common helper to use in IsReexecd(), ExecInSnapdOrCoreSnap() and here, but in all the cases, the helper was slightly awkward to use and kind of inconvenient. OTOH, I don't think we'll need more helpers for this, so maybe this one off thing is acceptable after all.

[the remark this answers is not in the document]
Comment (bboozzoo) [reply snapdtool/fips_linux.go:101]: So I sat down to it today again with clear head and came up with a helper that isn't too awkward to use. Please have a look.

File: snapdtool/fips_linux.go [hunk 8]
```diff
@@ -99,19 +99,20 @@ func MaybeSetupFIPS() error {
...
 	logger.Debugf("self exe: %s", exe)
+	logger.Debugf("exe root dir: %q", rootDir)
 
 	// on a classic system we need to be reexecuted from the snapd snap for
 	// the FIPS setup to be relevant, but on core we are not reexeced but
 	// running directly from the mount of the snapd snap under
 	// /usr/lib/snapd, yet we still want to set up the right environment
 	if release.OnClassic {
-		if !strings.HasPrefix(exe, currentRevSnapdSnap+"/") {
+		if !strings.HasPrefix(rootDir, dirs.SnapMountDir) {
```
Comment (pedronis) [inline snapdtool/fips_linux.go:115]: wouldn't at this point rootDir and  currentRevSnapdSnap be the same? do we need to compute  currentRevSnapdSnap?

**Review by pedronis — COMMENTED** [review]
thank you, question

File: snapdtool/fips_linux.go (hunk 2 above)
Comment (valentindavid) [reply snapdtool/fips_linux.go:131]: I see you added a comment before on the definition of GOFIPS=1. It looks fine. Thank you.

**Review by valentindavid — APPROVED** [review]

File: snapdtool/fips_linux.go (hunk 8 above)
Comment (bboozzoo) [reply snapdtool/fips_linux.go]: Ah yes, good point. Updated.

File: snapdtool/tool_test.go [hunk 9]
```diff
@@ -521,7 +521,10 @@ func (s *toolSuite) TestExeAndRoot(c *C) {
 	root, exe, err = snapdtool.ExeAndRoot()
 	c.Assert(err, IsNil)
 	c.Assert(root, Equals, dirs.GlobalRootDir)
-	c.Assert(exe, Equals, "usr/lib/snapd/snapd")
+	// dstro libexecdir without the root part
```
Comment (pedronis) [inline snapdtool/tool_test.go:524]: typo: distro

**Review by pedronis — APPROVED** [review]
thanks
